File: it-agents/skills/code-architect/references/code-examples/java/Observer.java
// Observer — subscription mechanism to notify subscribers of events.

import java.util.ArrayList;
import java.util.List;

interface Subscriber {
    void onEvent(String event);
}

class Publisher {
    private final List<Subscriber> subscribers = new ArrayList<>();
    void subscribe(Subscriber s) { subscribers.add(s); }
    void unsubscribe(Subscriber s) { subscribers.remove(s); }
    void publish(String event) {
        for (Subscriber s : subscribers) s.onEvent(event);
    }
}

class LoggingSubscriber implements Subscriber {
    private final String id;
    LoggingSubscriber(String id) { this.id = id; }
    public void onEvent(String event) {
        System.out.println(id + " handling event: " + event);
    }
}

public class Observer {
    public static void main(String[] args) {
        Publisher publisher = new Publisher();
        Subscriber a = new LoggingSubscriber("A");
        Subscriber b = new LoggingSubscriber("B");

        publisher.subscribe(a);
        publisher.subscribe(b);
        publisher.publish("created");

        publisher.unsubscribe(a);
        publisher.publish("updated");
    }
}

File: it-agents/skills/code-architect/references/code-examples/ruby/observer.rb
# frozen_string_literal: true

# Observer — subscription mechanism to notify subscribers of events.

# Subject maintains a list of observers and broadcasts changes.
class Subject
  def initialize = @observers = []

  def subscribe(observer)
    @observers << observer
    self
  end

  def unsubscribe(observer)
    @observers.delete(observer)
    self
  end

  def notify(event)
    @observers.each { |o| o.update(event) }
  end
end

# Concrete subject pushes an event whenever its value changes.
class Thermometer < Subject
  def reading=(value)
    @reading = value
    notify(degrees: value)
  end
end

# Observers react to events; any object with #update qualifies.
class Display
  def initialize(label) = @label = label
  def update(event) = puts "#{@label}: #{event[:degrees]}°"
end

if __FILE__ == $PROGRAM_NAME
  thermo = Thermometer.new
  phone = Display.new("phone")
  panel = Display.new("panel")

  thermo.subscribe(phone).subscribe(panel)
  thermo.reading = 21

  thermo.unsubscribe(panel)
  thermo.reading = 23
end
